core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -0,0 +1,469 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use bytes::BufMut;
+use bytes::Bytes;
+use tokio::io::ReadBuf;
+
+use std::cmp::min;
+use std::io::SeekFrom;
+
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use crate::raw::*;
+use crate::*;
+
+/// [BufferReader] allows the underlying reader to fetch data at the buffer's size
+/// and is used to amortize the IO's overhead.
+pub struct BufferReader<R> {
+    r: R,
+    cur: u64,
+
+    buf: Vec<u8>,
+    filled: usize,
+    pos: usize,
+}
+
+impl<R> BufferReader<R> {
+    /// Create a new [`oio::Reader`] with a buffer.
+    pub fn new(r: R, cap: usize) -> BufferReader<R> {
+        BufferReader {
+            r,
+            cur: 0,
+
+            buf: Vec::with_capacity(cap),
+            filled: 0,
+            pos: 0,
+        }
+    }
+
+    /// Invalidates all data in the internal buffer.
+    #[inline]
+    fn discard_buffer(&mut self) {
+        self.buf.clear();
+        self.pos = 0;
+        self.filled = 0;
+    }
+
+    /// Returns the capacity of the internal buffer.
+    fn capacity(&self) -> usize {
+        self.buf.capacity()
+    }
+}
+
+impl<R> BufferReader<R>
+where
+    R: oio::Read,
+{
+    fn poll_fill_buf(&mut self, cx: &mut Context<'_>) -> Poll<Result<&[u8]>> {
+        // If we've reached the end of our internal buffer then we need to fetch
+        // some more data from the underlying reader.
+        // Branch using `>=` instead of the more correct `==`
+        // to tell the compiler that the pos..cap slice is always valid.
+        if self.pos >= self.filled {
+            debug_assert!(self.pos == self.filled);
+
+            let cap = self.capacity();
+            self.buf.clear();
+            let dst = self.buf.spare_capacity_mut();
+            let mut buf = ReadBuf::uninit(dst);
+            unsafe { buf.assume_init(cap) };
+
+            let n = ready!(self.r.poll_read(cx, buf.initialized_mut()))?;
+            unsafe { self.buf.set_len(n) }
+
+            self.pos = 0;
+            self.filled = n;
+        }
+
+        Poll::Ready(Ok(&self.buf[self.pos..self.filled]))
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.pos = min(self.pos + amt, self.filled);
```

Comment:
When will `self.pos + amt > self.filled` happen? Do we need to handle the `self.cur` correct?